zipline/modelling/engine.py

```diff
@@ -0,0 +1,443 @@
+        # another parent.  Ensure that we have enough extra rows to satisfy
+        # all of our parents.
+        existing['extra_rows'] = max(extra_rows, existing['extra_rows'])
+
+    for subterm in term.inputs:
+        _add_to_graph(
+            subterm,
+            dependencies,
+            parents,
+            extra_rows=extra_rows + term.extra_input_rows,
+        )
+        dependencies.add_edge(subterm, term)
+
+    parents.remove(term)
+
+
+class FFCEngine(with_metaclass(ABCMeta)):
+
+    @abstractmethod
+    def factor_matrix(self, terms, start_date, end_date):
+        """
+        Compute the full factor matrix with a first date of `start`.
+
+        Parameters
+        ----------
+        terms : dict
+            Map from str -> zipline.modelling.term.Term.
+        start : datetime
+            The first date of the matrix.
+        end : datetime
+            The last date of the matrix.
+
+        Returns
+        -------
+        matrix : pd.DataFrame
+            A matrix of factors
+        """
+        raise NotImplementedError()
+
+
+class NoOpFFCEngine(FFCEngine):
+    """
+    FFCEngine that doesn't do anything.
+    """
+
+    def factor_matrix(self, terms, start, end):
+        return DataFrame(index=[], columns=sorted(terms.keys()))
+
+
+class SimpleFFCEngine(object):
+    """
+    FFC Engine class that computes each term independently.
+    """
+    __slots__ = [
+        '_loader',
+        '_calendar',
+        '_finder',
+        '__weakref__',
+    ]
+
+    def __init__(self, loader, calendar, asset_finder):
+        self._loader = loader
+        self._calendar = calendar
+        self._finder = asset_finder
+
+    def factor_matrix(self, terms, start_date, end_date):
+        """
+        Compute a factor matrix.
+
+        See Also
+        --------
+        FFCEngine.factor_matrix
+        """
+        graph = build_dependency_graph(terms.values())
+        ordered_terms = topological_sort(graph)
+        extra_row_counts = get_node_attributes(graph, 'extra_rows')
+        max_extra_rows = max(extra_row_counts.values())
+
+        lifetimes = self.build_lifetimes_matrix(
+            start_date,
+            end_date,
+            max_extra_rows,
+        )
+        lifetimes_between_dates = lifetimes[max_extra_rows:]
+
+        dates = lifetimes_between_dates.index.values
+        assets = lifetimes_between_dates.columns.values
+
+        raw_outputs = self.compute_chunk(
+            ordered_terms,
+            extra_row_counts,
+            lifetimes,
+        )
+
+        # We only need filters and factors to compute the final output matrix.
+        raw_filters = [lifetimes_between_dates.values]
+        raw_factors = []
+        factor_names = []
+        for name, term in iteritems(terms):
+            extra = extra_row_counts[term]
+            if isinstance(term, Factor):
+                factor_names.append(name)
+                # Remove extra rows.
+                if extra:
```

Comment:
(This is supposed to remove the `[extra:]` slice so we don't make an extraneous copy when extra is 0.)